src/helpers/getRandomWord.ts

```diff
@@ -1,35 +1,50 @@
-const shuffleArray = (array: string[]) => {
-  const newArray = [...array];
-  for (let i = newArray.length - 1; i > 0; i--) {
-    const j = Math.floor(Math.random() * (i + 1));
-    [newArray[i], newArray[j]] = [newArray[j], newArray[i]];
-  }
-  return newArray;
-};
 export const getRandomWord = (
   wordPack: readonly string[],
-  bannedWords: readonly string[]
+  bannedWords: readonly string[],
+  playedWords: Set<string>,
+  lastWord?: string | undefined
 ) => {
-  const randomIndex = Math.floor(Math.random() * wordPack.length);
-  const word = wordPack[randomIndex];
-  let scrabmledGoal = word;
-  let isSafe = false;
+  let availableWords = wordPack.filter((word) => !playedWords.has(word));
+
+  if (availableWords.length === 0) {
+    playedWords.clear();
+    if (lastWord) playedWords.add(lastWord);
+    availableWords = wordPack.filter((word) => !playedWords.has(word));
+  }
+
+  const randomIndex = Math.floor(Math.random() * availableWords.length);
+  const word = availableWords[randomIndex];
+
+  let scrambledGoal = word;
+  let isClean = false;
   let attempts = 0;
   const maxAttempts = 10;
 
-  while ((!isSafe || scrabmledGoal === word) && attempts < maxAttempts) {
-    scrabmledGoal = shuffleArray(word.split("")).join("");
+  while ((isClean || scrambledGoal === word) && attempts < maxAttempts) {
+    scrambledGoal = shuffleWord(word.split("")).join("");
 
-    // eslint-disable-next-line no-loop-func
-    isSafe = !bannedWords.some((banned) =>
-      scrabmledGoal.toUpperCase().includes(banned.toUpperCase())
-    );
+    isClean = isRandomWordClean(scrambledGoal, bannedWords);
 
     attempts++;
   }
 
   return {
     goal: word,
-    scrabmledGoal: scrabmledGoal,
+    scrambledGoal: scrambledGoal,
   };
 };
+
+const shuffleWord = (array: string[]) => {
```

Comment:
You can mark the argument as readonly since you're not modifying it.

```suggestion
const shuffleWord = (array: readonly string[]) => {
```